---
title: "Analysis and Formatting in R"
author: "Ashley Paulsen"
date: "December 4, 2018"
output: html_document
---
###File Guide

Microsatellite genotype data.xlsx has the raw microsatellite data

Haplotypes of mtDNA-final.txt has the raw haplotype data

final_project_AP.R has this code as a script, instead of a markdown file

microsatellite_geno_pop.txt has the genepop formatted microsatellite data

mtDNA_haplotypes.fasta has the FASTA formatted mitochondrial DNA

test_code.R has leftover bits of code that we didn't use, but are potentially useful
cmdline and fichier.in are files that genepop uses

ARPAR.xlsx has the allelic richness data from Hp-Rare

HFP_GenAlEx_Sheet.xlsx has the HFP sheet from GenAlEx

Table_2.csv has our remade version of Table 2 from the paper

###Importing and Formatting Data
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(genepop)
library(readxl)
library(tidyverse)
```

I had to go in excel and separate the sheets in the workbook into two different workbooks, microsatellite genotype data and sampling_information.

```{r Data Import}
microsatellite_genotype <- read_excel("Microsatellite genotype data.xlsx")
mtDNA_haplotypes <- read.table("Haplotypes of mtDNA-final.txt", sep = '\t', fill = T)
```

####FASTA formatting

We needed the mtDNA Haplotype information in FASTA fromat for another analysis.

```{r}
mtDNA_haplotypes[] <- lapply(mtDNA_haplotypes, as.character)
##save haplotype data as characters for grep

##formatting mtDNA_haplotypes into fasta format, by removing the location information
mtDNA_haplotypes <- subset(mtDNA_haplotypes, V3=="") 
mtDNA_haplotypes$V2 <- NULL
mtDNA_haplotypes$V3 <- NULL

##save the cleaned up tables
write.table(mtDNA_haplotypes, "mtDNA_haplotypes.fasta",quote = F, row.names = F,
            col.names = F)
```
####Table 2
Read in data
```{r}
AR_PAR <- read_excel("ARPAR.xlsx")
HFP <- read_excel("HFP_GenAlEx_Sheet.xlsx")
```

Trim and shape the data

```{r}
##Trim off empty/unneeded parts
AR_PAR <- AR_PAR[,c(1,2)]
HFP <- HFP[-c(1:112),]
##makes sure names are characters
HFP$`Heterozygosity, Fstatistics and Polymorphism by Population for Codominant Data` <- 
  as.character(HFP$`Heterozygosity, Fstatistics and Polymorphism by Population for Codominant Data`)
```

Make data frame for Table 2
```{r}
tbl2 <- data.frame("Locations" = c("Ning", "Sha", "Sui", "Xiang",
                                     "Huo", "DHS", "Hu", "Sand", "Nan",
                                     "Ding", "Wan", "Average"), 
                   "Na" = NA, "Ar"=NA, "PAr"=NA, "Ho"=NA, "He"=NA,
                   "Fis"=NA, stringsAsFactors = F)
```

Fill in the columns
Fill in Na column from the genalex data
```{r}
for (row in 1:34){
  m = HFP[row, c(1)]
  n = HFP[row, c(4)]
  if (is.na(m) == F){
    for (ro in 1:nrow(tbl2)) {
      p = tbl2[ro, c(1)]
      if (m == p){
        tbl2[ro, c(2)] = n
      }
    }
  }
}
```
Fill in Ho column from genalex
```{r}
for (row in 1:34){
  m = HFP[row, c(1)]
  n = HFP[row, c(7)]
  if (is.na(m) == F){
    for (ro in 1:nrow(tbl2)) {
      p = tbl2[ro, c(1)]
      if (m == p){
        tbl2[ro, c(5)] = n
      }
    }
  }
}
```

Fill in He column from genalex
```{r}
for (row in 1:34){
  m = HFP[row, c(1)]
  n = HFP[row, c(8)]
  if (is.na(m) == F){
    for (ro in 1:nrow(tbl2)) {
      p = tbl2[ro, c(1)]
      if (m == p){
        tbl2[ro, c(6)] = n
      }
    }
  }
}
```
Fill in F(Fis) column from genalex
```{r}
for (row in 1:34){
  m = HFP[row, c(1)]
  n = HFP[row, c(10)]
  if (is.na(m) == F){
    for (ro in 1:nrow(tbl2)) {
      p = tbl2[ro, c(1)]
      if (m == p){
        tbl2[ro, c(7)] = n
      }
    }
  }
}
```

Separate the allelic richness data into useable portions
```{r}
Ar <- AR_PAR[c(1:15),]
Ar <- Ar[-c(1,2, 14, 15),]

PAr <- AR_PAR[c(16:29),]
PAr <- PAr[-c(1:3),]
```

Merge by allelic richness tables and table 2 by row number, they must be in the same order to do so.
```{r}
for ( x in 1:12){
  tbl2[x, c(3)] = Ar[x, c(2)]
  tbl2[x, c(4)] = PAr[x, c(2)]
}
```

Convert all columns to numerics
```{r}
tbl2$Na = as.numeric(tbl2$Na)
tbl2$Ar = as.numeric(tbl2$Ar)
tbl2$PAr = as.numeric(tbl2$PAr)
tbl2$Ho = as.numeric(tbl2$Ho)
tbl2$He = as.numeric(tbl2$He)
tbl2$Fis = as.numeric(tbl2$Fis)
```

Find the average of each column
```{r}
for (d in 2:7) {
  tbl2[c(12), d] = mean(tbl2[c(1:11),d])
}
```

Write table 2
```{r}
write_excel_csv(tbl2,"Table_2.csv")
```


####Genepop Data Formatting

Genepop requires a specific file format.

```{r Data formatting}
##add commas on 2nd row
microsatellite_genotype[c(2),] =  paste(microsatellite_genotype[c(2),], ",", sep = "")
colnames(microsatellite_genotype) <- microsatellite_genotype[c(2),] ##rename columns
microsatellite_genotype <- microsatellite_genotype[-c(1, 2),] ##remove top two rows, now useless
microsatellite_genotype$`Individual ID,` =  paste(microsatellite_genotype$`Individual ID,`, ",")
  ##add commas to every ID
```

Genepop requires that every allele is identified by 2 number. This loop finds any position that has one char and adds a 0 in front so every position has 2 characters.

```{r}
microsatellite_genotype[] <- lapply(microsatellite_genotype, as.character)

for (col in 3:ncol(microsatellite_genotype)){
  for (row in 1:nrow(microsatellite_genotype)) {
    n <- microsatellite_genotype[row, col]
    if(nchar(n)!= 2) {
      n = paste("0", n, sep = "")
    }
    microsatellite_genotype[row, col] <- n
  }
}  
```

We have to combine the two columns for each locus, so every locus has 4 numbers, two number for each allele.
```{r}
for (n in 3:11){
  microsatellite_genotype[,c(n)] <- do.call(paste, c(microsatellite_genotype[,c(n,n+1)], sep = "")) 
  microsatellite_genotype <- microsatellite_genotype[,-c(n+1)]
}
```

Separate each population by the word pop, then remove the population column
```{r}
microsatellite_genotype <- add_row(microsatellite_genotype,`Individual ID,`='pop',
                                   `Population,` = 'NA', .before = 1)
for (row in 2:218) {
  x <- microsatellite_genotype[row, c(2)]
  y <- microsatellite_genotype[row+1, c(2)]
  if(x != "NA"){
    if(x != y){
      microsatellite_genotype <- add_row(microsatellite_genotype,
                                         `Individual ID,`='pop',`Population,` = 'NA',
                                         .after = row)
    }
  }
}
##remove population column
microsatellite_genotype$`Population,` <- NULL
```

We have to write it to a txt file and then edit the last bit by hand

Open microsatellite_geno_pop.txt and hit enter after "Individual ID," so that it is the first line and the next is the list of loci. Then remove the comma after SSR9(H33). Now we have genepop formatted data.

```{r}
write.table(microsatellite_genotype, "microsatellite_geno_pop.txt", na = "" ,quote = F,
            row.names = F, append = F)

```

###Genepop Analysis
####Hardy-Weinburg
```{}
test_HW("microsatellite_geno_pop.txt", which = "Proba", outputFile = "HW_microsatellite.txt.P", settingsFile = "",
        enumeration = FALSE, dememorization = 10000, batches = 100,iterations = 1000,
        verbose = interactive())
```

####Linkage Disequilibrium
```{}
test_LD("microsatellite_geno_pop.txt", outputFile = "LD_microsatellite.txt", settingsFile = "",
        dememorization = 10000, batches = 100, iterations = 5000,
        verbose = interactive())
write_LD_tables("microsatellite_geno_pop.txt", outputFile = "", verbose = interactive())
```

####FST
```{}
Fst("microsatellite_geno_pop.txt", sizes = FALSE, pairs = FALSE, outputFile = "FST_microsatellite.txt",
    dataType = "Diploid", verbose = interactive())
```





